File: cpp/creational/registry/nishith/Registry/Registry/src/InternRegistry.cpp
#include "InternRegistry.h"
#include <iostream>

void InternRegistry::Register(const std::string& key,
                              std::unique_ptr<Intern> prototype){
   registry_[key] = std::move(prototype);
}

std::unique_ptr<Intern> InternRegistry::create(const std::string& key) const {
   if (const auto it = registry_.find(key); it != registry_.end()) {
      return it->second->clone();
   }
   return nullptr;
}

File: java/creational/registry/uday/InternRegistry.java
import java.util.HashMap;
import java.util.Map;

/**
 * A registry for managing intern prototypes. Allows registering prototypes
 * with specific keys and creating new intern instances by cloning the
 * registered prototypes.
 */
public class InternRegistry {

    private final Map<String, Intern> registry = new HashMap<>();

    /**
     * Registers an intern prototype with a specific key.
     *
     * @param key       the key to associate with the prototype.
     * @param prototype the intern prototype to register.
     */
    public void register(String key, Intern prototype) {
        registry.put(key, prototype);
    }

    /**
     * Creates a new intern instance by cloning the prototype associated
     * with the specified key.
     *
     * @param key the key associated with the desired intern prototype.
     * @return a new intern instance, or {@code null} if no prototype is
     *         associated with the key.
     */
    public Intern create(String key) {
        Intern prototype = registry.get(key);
        if (prototype != null) {
            return prototype.clone1();
        }
        return null;
    }
}
